[GSoD]  Added tutorials directory on gRPC-Gateway (#1829)
grpc-ecosystem/grpc-gateway: docs/docs/tutorials/introduction.md

---
layout: default

> title: Introduction about gRPC-Gateway

Reviewer: ```suggestion
title: Introduction to the gRPC-Gateway
```

parent: Tutorials
nav_order: 1
---

## Introduction

> We all know that gRPC is not a tool for everything. There are cases where we still want to provide a traditional RESTful JSON API. The reasons can range from maintaining backwards-compatibility to supporting programming languages or clients not well supported by gRPC. But coding another API for REST is quite a time consuming and tedious.

Reviewer: Use `HTTP/JSON` to describe the JSON interface.
```suggestion
We all know that gRPC is not a tool for everything. There are cases where we still want to provide a traditional HTTP/JSON API. The reasons can range from maintaining backward-compatibility to supporting programming languages or clients not well supported by gRPC. But writing another service just to expose an HTTP/JSON API is quite a time consuming and tedious task.
```

> So is there any way to code just once, but can provide APIs in both gRPC and REST at the same time?

Reviewer: ```suggestion
So is there any way to code just once, but provide APIs in both gRPC and HTTP/JSON at the same time?
```

The answer is Yes.

The grpc-gateway is a plugin of the Google protocol buffers compiler [protoc](https://github.com/protocolbuffers/protobuf). It reads protobuf service definitions and generates a reverse-proxy server which translates a RESTful HTTP API into gRPC. This server is generated according to the [`google.api.http`](https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L46) annotations in your service definitions.

> This helps you provide your APIs in both gRPC and RESTful style at the same time.

Reviewer: ```suggestion
This helps you provide your APIs in both gRPC and HTTP/JSON format at the same time.
```

![architecture introduction diagram](https://docs.google.com/drawings/d/12hp4CPqrNPFhattL_cIoJptFvlAqm5wLQ0ggqI5mkCg/pub?w=749&h=370)

### Prerequisites

> Before start coding, we have to install some tools and need to do small amount of configuration in the service.

Reviewer: ```suggestion
Before we start coding, we have to install some tools and need to do a small amount of configuration in the service.
```

> Before we start coding, we have to install some tools and need to do a small amount of configuration in the service.

Reviewer: What service? Lets remove it.
```suggestion
Before we start coding, we have to install some tools.
```

We will be using a Go gRPC server in the examples, so please install Go first from [https://golang.org/dl/](https://golang.org/dl/).

After installing Go, use `go get` to download the following packages:

```sh
go get github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway
go get google.golang.org/protobuf/cmd/protoc-gen-go
go get google.golang.org/grpc/cmd/protoc-gen-go-grpc
```

> This installs the protoc generator plugins we need to generate the stubs. Make sure to add `$GOPATH/bin` to your `$PATH` so that executables installed via `go get` are available on your `$PATH`.

Reviewer: We should probably add a section below this about initializing a new module.
```suggestion
This installs the protoc generator plugins we need to generate the stubs. Make sure to add `$GOPATH/bin` to your `$PATH` so that executables installed via `go get` are available on your `$PATH`.

We will be working in a new module for this tutorial, so go ahead and create that in a folder of your choosing now:
''' (backticks)
$ go mod init github.com/myuser/myrepo
'''
```

Flesh it out a bit but you get the idea.

Reviewer: We also need to add a link to the next step in the tutorial series here. This goes for all the pages of the tutorial series.

parent: Tutorials
nav_order: 1
---

## Introduction

We all know that gRPC is not a tool for everything. There are cases where we still want to provide a traditional RESTful JSON API. The reasons can range from maintaining backwards-compatibility to supporting programming languages or clients not well supported by gRPC. But coding another API for REST is quite a time consuming and tedious.

So is there any way to code just once, but can provide APIs in both gRPC and REST at the same time?

The answer is Yes.

The grpc-gateway is a plugin of the Google protocol buffers compiler
[protoc](https://github.com/protocolbuffers/protobuf).
It reads protobuf service definitions and generates a reverse-proxy server which
translates a RESTful HTTP API into gRPC. This server is generated according to the
[`google.api.http`](https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L46)
annotations in your service definitions.

This helps you provide your APIs in both gRPC and RESTful style at the same time.

![architecture introduction diagram](https://docs.google.com/drawings/d/12hp4CPqrNPFhattL_cIoJptFvlAqm5wLQ0ggqI5mkCg/pub?w=749&h=370)

## Prerequisites

> One way to achieve that is to use gRPC gateway. gRPC gateway is a plugin of the protocol buffer compiler. It reads the protobuf service definitions and generates a proxy server, which translates a RESTful HTTP call into gRPC request.

Reviewer: Remove this sentence now, it doesn't flow well with the previous content.
